lib/internal/modules/esm/fetch_module.js

```diff
@@ -127,7 +138,8 @@ function fetchWithRedirects(parsed) {
         err.message = `Cannot find module '${parsed.href}', HTTP 404`;
         throw err;
       }
-      if (res.statusCode > 303 || res.statusCode < 200) {
+      if ((res.statusCode > 303 && !isRedirect(res.statusCode)) ||
+          res.statusCode < 200) {
```

Comment:
The original code was intending to catch all status codes in the 1xx, 4xx and 5xx ranges. The author didn’t realize that there were valid redirect codes higher than 303; a better version would just start at 400:

```suggestion
      if (res.statusCode < 200 || res.statusCode >= 400) {
```

Reply:
Thanks for explaining, this code always looks awkward to me, but I did not change it in order to be cautious.

Reply:
I'd like to point out that `3xx` codes are not only redirects: for example, there is `304 Not Modified`.

Reply:
Sorry, I should correct myself: that was my assumption as to what the original code was intending. @JakobJingleheimer perhaps you can confirm? Should this condition catch 304 and/or any of the other 3xx codes?

Reply:
I don't know why Bradley did not include them (he wrote this part). Perhaps it has to do with the subtle differences/specifics in 307 and 308?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#redirection_messages

> **`307`**
> This has the same semantics as the `302` Found HTTP response code, with the exception that the user agent must not change the HTTP method used: if a `POST` was used in the first request, a `POST` must be used in the second request.

(308 is the same, but relative to 301)

I'd need to check through the other code to see if it even supports changing the method during a redirect—I don't recall it supporting that. If it can't, I see no reason to not include 307 and 308.

Reply:
The nuances between 301/302 and 307/308 is unrelated here since it is only doing GET. But upon taking a close look I feel this is throwing a wrong error. If the status code is < 200 or >= 400 (or > 303 as originally written), why would the code expect it to have a `location` header?